src/mame/namco/ns10crypt.h

```diff
@@ -157,6 +157,51 @@ class startrgn_decrypter_device : public ns10_type2_decrypter_device
 	static const ns10_crypto_logic crypto_logic;
 };
 
+class gahaha_decrypter_device : public ns10_type2_decrypter_device
+{
+public:
+	gahaha_decrypter_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+private:
+	static uint16_t nonlinear_calc(uint64_t previous_cipherwords, uint64_t previous_plainwords, const gf2_reducer &reducer);
+	static const ns10_crypto_logic crypto_logic;
+};
+
+class g13jnc_decrypter_device : public ns10_type2_decrypter_device
+{
+public:
+	g13jnc_decrypter_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock);
+private:
+	static uint16_t nonlinear_calc(uint64_t previous_cipherwords, uint64_t previous_plainwords, const gf2_reducer &reducer);
+	static const ns10_crypto_logic crypto_logic;
+};
```

Comment:
Not really critical at the moment, but if the only difference between these devices a reference to a structure and a function, they could be reduced to a single device with the structure and function (as a lambda) supplied in machine configuration of the client.

Reply:
I'll see about writing it to work in the machine configuration. I have a feeling that would be easier to maintain too because right now you have to touch a lot of places just to add a new decrypter.

Reply:
Done. Reduced down to just `NS10_TYPE2_DECRYPTER` and `MRDRILR2_DECRYPTER` (only known Type 1 game). Tested all games with decrypters to make sure everything is still booting.